### PR TITLE
[dtensor][5/n] switch random ops to op strategy

### PR DESCRIPTION
**Pull request by wanchaol**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #109145
* #109144
* #108264
* __->__ #108263

This PR switches the random ops to use op strategy instead of rule
based, this is a first series of PRs to refactor ops after we refactor
op dispatch logic

### CODE REVIEW

File: torch/distributed/_tensor/sharding_prop.py [hunk 1]
```diff
@@ -170,8 +172,14 @@ def spec_to_strategy(spec: object) -> object:
                 reshard_schema._inplace_rewrap_schema_suggestion(op_schema)
                 suggestion_schema = [reshard_schema]
 
+            output_spec: OutputSpecType = output_strategy.output_spec
+            if op_schema.op == aten.native_dropout.default:
+                # for ops return multiple tensors, make output spec return same spec
+                # returned from the op strategy
+                output_spec = (output_strategy.output_spec, output_strategy.output_spec)
```
Comment (fduwjj) [inline torch/distributed/_tensor/sharding_prop.py:179]: Can we have this in the random_op_strategy?

**Comment by fduwjj** [comment]
Also another n00b question, does this mean we want to move all rules, ideally, to strategies? Because keeping two ways of registering op seems to be a little bit confusing to users who want to register their own op?

File: torch/distributed/_tensor/sharding_prop.py (hunk 1 above)
Comment (wanchaol) [reply torch/distributed/_tensor/sharding_prop.py:179]: Just tried, but it's a bit non-trival to make the `output_spec` to be a tuple of dtensor spec and resharding becomes complicated. I'll try to refactor this in a more canonical way

**Comment by wanchaol** [comment]
> Also another n00b question, does this mean we want to move all rules, ideally, to strategies? Because keeping two ways of registering op seems to be a little bit confusing to users who want to register their own op?

Yep! the north star is that all ops going forward can be implemented with strategy instead of rule, that would make our stack cleaner with possible easier op implementation

**Review by fduwjj — APPROVED** [review]
Unblock for now, and looking forward to more refactoring.

**Pull request closed without merging**